File: src/neurnet.erl
%%%-------------------------------------------------------------------
%%% @author borja
%%% @doc
%%%
%%% TODO: Un agente debe de estar formado por varias redes neuronales de control y la superior
%%% TODO: Añadir metodo de ajuste de pesos mediante estimulos positivos/negativos
%%% TODO: Debe de haber una population id "Real" que siempre existe, si no se especifica el pop_id, el agente se refiere ahu
%%% @end
%%%-------------------------------------------------------------------
-module(neurnet).
-author("borja").
-compile([export_all, nowarn_export_all]). %%TODO: To delete after build

%% API
-export([]).


%%%===================================================================
%%% API
%%%===================================================================

%%--------------------------------------------------------------------
%% @doc Returns a list of tuples with the record name and attributes
%% list. This is mainly used to prepare the tables in mnesia.
%% @end
%%--------------------------------------------------------------------
-spec attributes_table() -> 
    [{Elem :: nnelements:element(), [Attr :: atom()]}].
attributes_table() ->
    eevo:attributes_table() ++ enn:attributes_table() ++ [
        % Genotype
        % Phenotype
        % Architectures

        {      actuator, actuator:fields(      actuator)},
        {actuator_group, actuator:fields(actuator_group)},
        {        sensor,   sensor:fields(        sensor)},
        {  sensor_group,   sensor:fields(  sensor_group)}
    ].

%%--------------------------------------------------------------------
%% @doc Loads a module with the genotypes to test in neurnet. 
%% @end
%%--------------------------------------------------------------------
-spec load(GenotypeModules :: [module()]) -> ok.
load(GenotypeModules) -> 
    genotype:load(GenotypeModules).

%%--------------------------------------------------------------------
%% @doc Triggers a genotype per population to evaluate an evolutionary
%% algorithm with neural networks and returns the score and the id of
%% the best phenotype for each specified genotype.
%% @end
%%--------------------------------------------------------------------
%TODO: 


%%--------------------------------------------------------------------
%% @doc 
%% @end
%%--------------------------------------------------------------------
%TODO: Correct specs
phenotype(Genotype) -> 
    Phenotype_Id = genotype:buid(Genotype),
    eevo:agent(#{
        function  => fun phenotype:controller/1,
        mutation  => fun phenotype:mutator/1,
        arguments => Phenotype_Id
    }).


%%%===================================================================
%%% Internal functions
%%%===================================================================

File: lib/neurnet.ex
defmodule Neurnet do
  @moduledoc """
  Documentation for `Neurnet`.
  """

  @type genotype() :: atom()
  @type phenotype() :: :eevo.agent()

  ### =================================================================
  ###  API
  ### =================================================================

  @doc """
  Loads the actuators from the indicated modules.
  """
  @spec load(modules :: [module :: atom]) :: {:atomic, [Genotype]}
  def load(modules) do
    Database.run(fn -> Genotype.load(modules) end)
  end

  @doc """
  Creates a phenotype from a genotype.
  """
  @spec phenotype_from(genotype()) :: phenotype()
  def phenotype_from(name) do
    genotype = Database.dirty_read!({:genotype, name})

    :eevo.agent(%{
      function: &Phenotype.controller/1,
      mutation: &Phenotype.mutate/1,
      arguments: Phenotype.from(genotype)
    })
  end

  @doc """
  Returns the phenotype information.
  """
  @spec info(phenotype()) :: term()
  def info(agent) do
    {:atomic, info} =
      :mnesia.transaction(fn ->
        %{arguments: [phenotype]} = :eevo.info(agent)

        %{
          actuators: phenotype.actuators,
          sensors: phenotype.sensors,
          network: %{
            id: phenotype.network,
            info: :nnet.info(phenotype.network)
          }
        }
      end)

    info
  end

  @spec architecture(phenotype()) :: :ok
  def architecture(agent) do
    {:atomic, architecture} =
      :mnesia.transaction(fn ->
        %{arguments: [phenotype]} = :eevo.info(agent)
        :umlnn.format(phenotype.network)
      end)

    :io.format(architecture)
  end

  @doc """
  Runs a phenotypes out of a population
  """
  @spec run_as(phenotype()) :: {:ok, pid()}
  def run_as(phenotype) do
    :eevo.run_as(phenotype)
  end

  @doc """
  Runs a training evolutionary algorithm with phenotypes
  """
  @spec run(atom(), genotype(), number(), function()) :: :eevo.results()
  def run(name, genotype, parallel, stop_condition) do
    phenotypes = for _ <- 1..parallel, do: phenotype_from(genotype)
    population = :eevo.population(name, :ramp3)
    :eevo.run(population, phenotypes, parallel, stop_condition)
  end

  ### =================================================================
  ###  Internal functions
  ### =================================================================
end
